### ISSUE TITLE
`is_a` allows for a variant check but it gives incorrect results

### ISSUE THREAD

**Issue by radeusgd** [enso-org/enso#5792]
Apply the following patch to `Meta_Spec.enso` and run it:
```diff
Index: test/Tests/src/Semantic/Meta_Spec.enso
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/test/Tests/src/Semantic/Meta_Spec.enso b/test/Tests/src/Semantic/Meta_Spec.enso
--- a/test/Tests/src/Semantic/Meta_Spec.enso	(revision bb90a92a25ff81ae9440e3f9899509c1feb95630)
+++ b/test/Tests/src/Semantic/Meta_Spec.enso	(date 1677755709960)
@@ -38,6 +38,10 @@
 type Test_Type
     Value x
 
+type Sum_Type
+    Variant_A x
+    Variant_B y
+
 spec =
     Test.group "Meta-Value Manipulation" <|
         Test.specify "should allow manipulating unresolved symbols" <|
@@ -125,6 +129,10 @@
             Meta.is_a Date_Time.now.zone Time_Zone . should_be_true
             Meta.is_a Date_Time.now.zone Date . should_be_false
 
+            (Sum_Type.Variant_A 42).is_a Sum_Type . should_be_true
+            (Sum_Type.Variant_A 42).is_a Sum_Type.Variant_A . should_be_true
+            (Sum_Type.Variant_A 42).is_a Sum_Type.Variant_B . should_be_false
+
         Test.specify "should allow for returning the type of a value" <|
             n_1 = Meta.type_of 42
             n_1 . should_equal_type Integer
```

The test will fail at the last line - apparently currently our `is_a` will answer YES to a question if `Variant_A` is a `Variant_B`. That is bad.

We can see that in `IsValueOfTypeNode` we have `doAtom` specialization which gets the parent type of the specific constructor and does the check on that. I.e. an `x.is_a Sum_Type.Variant_B` will essentially get converted to a check like `x.is_a Sum_Type`. But these two conditions are not at all equivalent!

Ideally, we want this to behave as in the test mentioned above - we'd like to be able to check a particular constructor using `is_a`, but it should work only for the specific variant.

If for some reason we cannot get that, or if we want a different approach for consistency, I think we could restrict the `is_a` check to just _types_ and not _constructors_ on the right. I.e. making `x.is_a Sum_Type.Variant_A` invalid, preferably throwing sth like `Illegal_Argument.Error`. However, we still are relying on checks like that in the test library so ideally we would want some alternative to perform these, like `Meta.is_atom_variant` which would do:
```
Meta.is_atom_variant Sum_Type.Variant_A (Sum_Type.Variant_A 42) == True
Meta.is_atom_variant Sum_Type.Variant_B (Sum_Type.Variant_A 42) == False
```

**Comment by radeusgd** [comment]
I think this may have relatively high priority because many of our tests rely on `is_a` (indirectly, by using `should_be_a` and `should_fail_with` which in turn use `is_a` internally).

Thus, our tests are not really testing what we think they are! They are testing that the parent type is right, but in many places we check for a particular variant and think that are testing that it is right but in fact that is not checked.

**Comment by radeusgd** [comment]
If we decide to fix `is_a` to the above mentioned spec, I think we should also add these checks to verify that this also works correctly for types where the constructor has no parameters (that is because then it may get instantiated immediately and `is_a` may get an `Atom` and not `AtomConstructor`).

```diff
Index: test/Tests/src/Semantic/Meta_Spec.enso
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/test/Tests/src/Semantic/Meta_Spec.enso b/test/Tests/src/Semantic/Meta_Spec.enso
--- a/test/Tests/src/Semantic/Meta_Spec.enso	(revision bb90a92a25ff81ae9440e3f9899509c1feb95630)
+++ b/test/Tests/src/Semantic/Meta_Spec.enso	(date 1677756302855)
@@ -38,6 +38,10 @@
 type Test_Type
     Value x
 
+type Parameterless_Sum_Type
+    type Variant_A
+    type Variant_B
+
 spec =
     Test.group "Meta-Value Manipulation" <|
         Test.specify "should allow manipulating unresolved symbols" <|
@@ -125,6 +129,10 @@
             Meta.is_a Date_Time.now.zone Time_Zone . should_be_true
             Meta.is_a Date_Time.now.zone Date . should_be_false
 
+            Parameterless_Sum_Type.Variant_A.is_a Parameterless_Sum_Type . should_be_true
+            Parameterless_Sum_Type.Variant_A.is_a Parameterless_Sum_Type.Variant_A . should_be_true
+            Parameterless_Sum_Type.Variant_A.is_a Parameterless_Sum_Type.Variant_B . should_be_false
+
         Test.specify "should allow for returning the type of a value" <|
             n_1 = Meta.type_of 42
             n_1 . should_equal_type Integer
```

**Comment by JaroslavTulach** [comment]
As pointed out in [pivotal](https://www.pivotaltracker.com/story/show/183970789) the previous behavior was [somewhat weird](https://github.com/enso-org/enso/pull/3949#discussion_r1040058599). The current behaviour is a result of aligning `Meta.is_a T` with case `x : T` check.  We have decided they should work the same. As case `x : T` doesn't work for constructors, so `Meta.is_a` shouldn't.

- Speed [Count/Min/Max benchmark](https://www.pivotaltracker.com/story/show/182124856) by not using `is_a`.
- Code [change](https://github.com/enso-org/enso/commit/011d9b81a1693153bb6a19339eb453d333ac698c) on branch
- Next [pivotal](https://www.pivotaltracker.com/story/show/183918642)
- Speed [Meta.is_a up](https://www.pivotaltracker.com/story/show/183970789)
- #3949

@radeusgd had expressed his [positive attitude towards constructor check](https://github.com/enso-org/enso/pull/3949#discussion_r1041288373), but he also noted:

> not sure if is_a is it's right place - I still think that ensuring is_a is aligned with case-of on types is the way to go

E.g. I believe the current implementation of `Meta.is_a` is correct and we need to find another place to do constructor/variant checks.

**Comment by radeusgd** [comment]
> E.g. I believe the current implementation of `Meta.is_a` is correct

I think it is close to correct, but actually has a bug which is demonstrated above.

In no circumstances I can see `(Sum_Type.Variant_A 42).is_a Sum_Type.Variant_B == True` being the expected behaviour. It is confusing as it suggests that A is B which it isn't. If we want to keep `is_a` consistent with `_ : T`, we should simply disallow non-type values on the right-hand-side. So IMO `x.is_a Type.Constructor` should ideally raise some `Illegal_Argument` error.

> and we need to find another place to do constructor/variant checks.

I think we can live without it, as long as `is_a` is fixed to be correct - so that we get rid of the false sense of correctness our current tests give (we have checks like `foo.should_fail_with File_Error.IO_Error` which actually only check if `foo` is `File_Error` but do not check if it's `IO_Error` or `Not_Found` or sth else - and that is bad because the test looks like it is checking something but it is not. Once `is_a` is fixed to actualy be consistent with `_ : T`, these tests will simply start failing (but that's a good sign because they will no longer give us false information) - so we will need to fix these somehow.

**Comment by radeusgd** [comment]
I managed to create a 'workaround' alternative for `is_a` using existing tooling that will mallow us to match constructors:

```
from Standard.Base import all
import Standard.Base.Runtime.Debug

Any.adapted_is_a self thing =
    m = Meta.meta thing
    fallback =
        self.is_a thing
    case m of
        # Regular ctor case
        Meta.Constructor.Value _ ->
            if Meta.is_atom self . not then False else
                Meta.meta self . constructor == m
        # Parameterless ctor case (it will already be applied so we un-apply it)
        Meta.Atom.Value _ ->
            is_parameterless_constructor = m.constructor.fields.is_empty
            if is_parameterless_constructor.not then fallback else
                Meta.meta self . constructor == m.constructor
        _ -> fallback

type Sum_Type
    Variant_A x
    Variant_B y

type Parameterless_Sum_Type
    Variant_A
    Variant_B

main =
    # The first line does not work due to #5805 bug.
    #IO.println (Sum_Type.Variant_A 1 . adapted_is_a Sum_Type)
    IO.println (Sum_Type.Variant_A 1 . adapted_is_a Sum_Type.Variant_A)
    IO.println (Sum_Type.Variant_A 1 . adapted_is_a Sum_Type.Variant_B)

    #IO.println (Parameterless_Sum_Type.Variant_A . adapted_is_a Parameterless_Sum_Type)
    IO.println (Parameterless_Sum_Type.Variant_A . adapted_is_a Parameterless_Sum_Type.Variant_A)
    IO.println (Parameterless_Sum_Type.Variant_A . adapted_is_a Parameterless_Sum_Type.Variant_B)
```

So I would suggest to make this ticket to fix `is_a` to _not_ allow the faulty behaviour where `Sum_Type.Variant_A . is_a Sum_Type.Variant_B == True` (cc: @JaroslavTulach). And the other part can probably be handled by us in the library team.

@jdunkerley with the snippet above we can fix the constructor matching. I think we could move the part of this logic to a helper `Meta.is_from_constructor object constructor` (name TODO). We can then amend `should_fail_with` and `should_be_a` to use similar logic to the snippet above to determine if it should call `is_from_constructor`, `is_a` or just `==`. Shall I create a task to do it?

**Comment by jdunkerley** [comment]
Yes agree with `is_a` being type only. In which case, it should error if passed a non-type argument.
We need a ` Meta.is_built_from` (and then extensions as we wish) for constructors.

`Panic.catch` and `Error.catch` should probably then only work with types as well.

We will need to amend and update the test suite to allow us to test for either.

**Comment by radeusgd** [comment]
So to sum up, we will need the following test suite to pass:

```
type Sum_Type
    Variant_A x
    Variant_B y

type Parameterless_Sum_Type
    Variant_A
    Variant_B

v1 = Sum_Type.Variant_A 42
v2 = Parameterless_Sum_Type.Variant_A

v1.is_a Sum_Type . should_be_true
v1.is_a Sum_Type.Variant_A . should_fail_with Illegal_Argument
v1.is_a Sum_Type.Variant_B . should_fail_with Illegal_Argument

v2.is_a Parameterless_Sum_Type. should_be_true
v2.is_a Parameterless_Sum_Type.Variant_A . should_fail_with Illegal_Argument

v1.is_built_from Sum_Type.Variant_A . should_be_true
v1.is_built_from Sum_Type.Variant_B . should_be_false

v2.is_built_from Parameterless_Sum_Type.Variant_A . should_be_true
v2.is_built_from Parameterless_Sum_Type.Variant_B . should_be_false
```

and we then want to amend in our `Test` library the methods `should_be_a` and `should_fail_with` to do something like `matches value matcher` where:
```
matches value matcher =
    if matcher is a constructor then
        value.is_built_from matcher
    else if matcher is a Type then
        value.is_a matcher
    else
        value == matcher
```

Ideally we should be able to do the `matcher is a Type` thanks to variant B of #5805.
We can do `matcher is a constructor` as I shown above with `Meta.meta` with the special case for 0-argument constructors.

I assume all of this can be part of this single ticket since they are all very related - essentially we need to fix `is_a` and split the constructor logic that we thought works but apparently it doesn't into `is_built_from` and then amend these 2 methods in the test suite that relied on that logic to now work correctly.

**Comment by JaroslavTulach** [comment]
OK. Thanks guys, I'll look into it soon.

**Comment by radeusgd** [comment]
Note: As we discussed with James on catchup, the specific `matches` logic that I wrote above for `should_fail_with` and `should_be_a` also applies to `Panic.catch` and `Error.catch`, well maybe without the `==` part. But still we may want to have the logic
```
matches value matcher = case matcher of
    ?Type? -> value.is_a matcher
    ?Constructor? -> value.is_built_from matcher
```
to be encapsulated in some helper function in `Meta` to be able to use it across places.

Highlevel, we want `Panic.catch` and `Error.catch` to work both with Types (and Java Classes too!) and Constructors, so that we can ofc keep doing as we are `Panic.catch JavaException ...` but can also do:
```
Panic.catch File_Error.IO_Error action handler
```
and make it only catch `IO_Error`s but **not** other variants of `File_Error`. We have a few of these multi-variant error types and it is useful to be able to catch a specific variant.

I think it would be good to add a few tests for this variant-matching in `Panic.catch` too, as part of this ticket.
